source/agora/test/Ledger.d

```diff
@@ -39,11 +40,7 @@ unittest
 
     foreach (idx; 0 .. 100)
     {
-        Transaction tx =
-        {
-            [Input(hashFull("Message No. %s".format(idx + 1)))],
-            [Output(100)]
-        };
+        Transaction tx = newCoinbaseTX(KeyPair.random().address);
```

Comment:
This is not really correct code. There are 100 coinbase transactions being made here. Would it make sense that a user could submit a new coinbase transaction at any point in time, even though only one should exist?

There needs to be a validation step to actually reject this test-case.

To actually fix the unittest, you could write something like this:

```diff
diff --git a/source/agora/common/Block.d b/source/agora/common/Block.d
index ea64945..a868697 100644
--- a/source/agora/common/Block.d
+++ b/source/agora/common/Block.d
@@ -115,10 +115,7 @@ public Block getGenesisBlock ()
     auto address = KeyPair.fromSeed(
         Seed.fromString("SCT4KKJNYLTQO4TVDPVJQZEONTVVW66YLRWAINWI3FZDY7U4JS4JJEI4")).address;
 
-    auto gen_tx = Transaction(
-        [Input(Hash.init, 0, Hash.init)],
-        [Output(40_000_000, address)]
-    );
+    auto gen_tx = newCoinbaseTX(address, 40_000_000);
 
     auto header = BlockHeader(
         Hash.init, 0,
diff --git a/source/agora/test/Ledger.d b/source/agora/test/Ledger.d
index 26b8cd1..07755f3 100644
--- a/source/agora/test/Ledger.d
+++ b/source/agora/test/Ledger.d
@@ -15,6 +15,7 @@ module agora.test.Ledger;
 
 version (unittest):
 
+import agora.common.crypto.Key;
 import agora.common.Block;
 import agora.common.Data;
 import agora.common.Hash;
@@ -37,16 +38,29 @@ unittest
     auto node_1 = network.apis.values[0];
     Transaction[] txes;
 
+    // same key-pair as in getGenesisBlock()
+    auto key_pair = KeyPair.fromSeed(
+        Seed.fromString("SCT4KKJNYLTQO4TVDPVJQZEONTVVW66YLRWAINWI3FZDY7U4JS4JJEI4"));
+
+    auto gen_block = getGenesisBlock();
+
+    // last transaction in the ledger
+    Hash last_tx_hash = hashFull(gen_block.tx);
+
     foreach (idx; 0 .. 100)
     {
         Transaction tx =
         {
-            [Input(hashFull("Message No. %s".format(idx + 1)))],
-            [Output(100)]
+            [Input(last_tx_hash, 0)],
+            [Output(40_000_000, key_pair.address)]  // send to the same address
         };
 
+        auto signature = key_pair.secret.sign(hashFull(tx)[]);
+        tx.inputs[0].signature = signature;
+
         node_1.putTransaction(tx);
         txes ~= tx;
+        last_tx_hash = hashFull(tx);  // we need to reference it again
     }
 
     // ensure block height is the same everywhere
```

To simplify:

GenesisTransaction => Transaction 2 (is spending unspent amount in GenesisTransaction) => Transaction 3 (is spending unspent amount in Transaction 2) => and so on..

Reply:
First of all, thank you for the good explanation.
So I did it to make the blocks simple.
I will not use coinBase in the test, and I will refer to the sample presented.